Model flattening for code generation.

hsorby opened the pull request:
Addresses #386.

hsorby committed:
Add in initial flattening test.

hsorby committed:
Merge branch 'develop' into model_flattening_i386

hsorby committed:
Clear previous errors when a new model is parsed by the parser.

hsorby committed:
Add basic tests for flattening a model with a units import and a component import.

hsorby committed:
Impement a method flatten on the Model class.

hsorby committed:
Change indentation for math to allow for what libXml2 does.

hsorby committed:
Change the expectation for parsing an empty model.

hsorby committed:
Give the expected response for attempting to parse an empty model.

hsorby committed:
Functionality to allow for importing components with internally mapped variables.

hsorby committed:
Add test for imported components that make connections outside of the imported component.

hsorby committed:
Add test for checking units used by variables in imported model are added to importing model.

hsorby committed:
Copy over units used by variables in imported model to imorting model.

hsorby committed:
Add a test for several layers of imported components.

hsorby committed:
Implement code for handling multiple layers of imports.

hsorby committed:
Add documentation for hasImports().

hsorby committed:
Add test for case where an imported component doesn't exist in the imported model.

hsorby committed:
Add test where imported components make use of the same units in the imported model.

hsorby committed:
Only add imported units to flattened model if they don't already exist in the flattened model.

hsorby committed:
Add test for importing units the same as units already in the importing model.

hsorby committed:
Merge branch 'develop' into model_flattening_i386

hsorby committed:
Add hasUnlinkedUnits() and linkUnits() to Model class.

hsorby committed:
Add check for valid standard prefixes in updateUnitMultiplier().

hsorby committed:
Add new line to printing out each error in a Logger in the utility printErrors function..

hsorby committed:
Use model->linkUnits() to link units in the isolated generator test.

hsorby committed:
Make the units 'common_units' dimensionally different so forcing a name change for the imported units.

hsorby committed:
Make use of linkUnits() to track used units in imported components.  Make units linked when parsing models with the parser.

hsorby committed:
Add file src/debug.cpp.

hsorby committed:
Create an internaltypes.h header file.  Add printStack and printEquivalenceMap to debug.cpp.

hsorby committed:
Remove creation list of units names in component hierarchy search.

hsorby committed:
Remove debugging code from model.cpp.

hsorby committed:
Remove method that is not used on private model class.

hsorby committed:
Add test for importing components that have name clashes.

hsorby committed:
Format the code according to clang format.

hsorby committed:
Attempt at adding a HH model that uses imports, unfortunately the variable ordering changes making an easy comparison not possible.

hsorby committed:
Replace 'vector.size() > 0' with '\!vector.empty()'.

hsorby committed:
Pass parameters as const referernces for smart pointers in flatten* functions.

hsorby committed:
Convert iterator based for loops to range based for loops.

hsorby committed:
Use int64_t instead of long.

hsorby committed:
Replace 'input.size() == 0' with '\!input.empty()' in parser.cpp.

hsorby committed:
Replace SIZE_T_MAX with SIZE_MAX.

hsorby committed:
Remove redundant string initialisations.

hsorby committed:
Add Python documentation for new Model API.

hsorby committed:
Add layered component tests for unlinked units.

hsorby committed:
Modify HH import model to manipulate code generation so that it is the same as HH model that doesn't use imports.

hsorby commented:
Ready for review.

nickerso committed:
Tweaking flatten method docs.

nickerso committed:
Make flatten docs consistent

nickerso commented on src/model.cpp:
```diff
@@ -221,6 +222,78 @@ size_t Model::unitsCount() const
+    }
+}
+
+void traversComponentTreeLinkingUnits(const ComponentPtr &component)
```
I want to add an 'e' to the `travers`...

agarny replied:
Done.

nickerso commented on src/model.cpp:
```diff
@@ -221,6 +222,78 @@ size_t Model::unitsCount() const
+    return unlinked;
+}
+
+bool traveseComponentTreeForUnlinkedUnits(const ComponentPtr &component)
```
ok, now I'm starting to think these are typos that were not picked up in code-completion tools :)

agarny replied:
Fixed too.

nickerso commented on tests/model/model_flattening.cpp:
```diff
@@ -0,0 +1,623 @@
+
+TEST(ModelFlattening, importedUnits)
+{
+    const std::string unitsDefiniitonModel =
```
these typos should probably be fixed, but will be easier to do them all at once in a proper editor rather than online via Github.com :)

agarny replied:
Done.

nickerso committed:
Merge branch 'develop' into model_flattening_i386

agarny commented:
Not sure whether @hsorby is back to work, but before I can review this, a conflict should be resolved. I am happy to do it, if needed. @nickerso?

nickerso commented:
go for it @agarny - thanks!

agarny committed:
Merge remote-tracking branch 'upstream/develop' into model_flattening_i386.

agarny committed:
Some minor cleaning up.

agarny commented on tests/model/model_flattening.cpp:
```diff
@@ -0,0 +1,623 @@
+    auto modelNonImportVersion = parser->parseModel(fileContents("generator/hodgkin_huxley_squid_axon_model_1952/model.cellml"));
+    auto e = printer->printModel(modelNonImportVersion);
+    // Sadly cannot get the order of the connections to match each other, everything else does.
+    // EXPECT_EQ(e, a);
```
I am going to reenable that check since we clearly want it. I am just going to double check the "original" HH52 model and modify accordingly, if it is acceptable.

agarny replied:
Done.

agarny commented on tests/model/model_flattening.cpp:
```diff
@@ -0,0 +1,623 @@
+        "      <ci>x</ci>\n"
+        "    </apply>\n"
+        "  </apply>\n"
+        "</math>\n"
```
This is not "nice". I would expect the indentation to be right. Will try to "fix" it.

agarny replied:
Done.

agarny commented on tests/units/units.cpp:
```diff
@@ -345,6 +345,173 @@ TEST(Units, replaceUnits)
+    m->addUnits(u1);
+
+    EXPECT_TRUE(m->hasUnlinkedUnits());
+}
```
Not sure about the need of this test. It's the same as `hasUnlinkedUnitsWhenBaseUnitsAddedToModelAfterNameUsedForVariableUnits` except that we add `second` to `a_unit`, but why should that make things different when it comes to `hasUnlinkedUnits()`? What am I missing?...

nickerso replied:
Isn't it to test that there is no special treatment of base units that is affecting this behaviour? (as we've found out in other aspects of units)

agarny replied:
Hmm... not convinced. We have `v1->setUnits("a_unit")`, which in itself results in an unlinked unit since `u1` is only added to the model after setting the units to `v1`. Anyway, I am not going to have this hold back the PR. I just thought I would mention my "concerns" so that @hsorby can look into it when he's back.

Right now, the only thing that I want before accepting this PR is to fix the indentation of the math, i.e. https://github.com/cellml/libcellml/pull/483#discussion_r363150645.

agarny commented on tests/units/units.cpp:
```diff
@@ -345,6 +345,173 @@ TEST(Units, replaceUnits)
+    v3->setUnits(u1);
+
+    EXPECT_TRUE(m->hasUnlinkedUnits());
+}
```
Again, I am not quite sure why the `hasUnlinkedUnitsLayeredComponentsWithVariables` and `hasUnlinkedUnitsSiblingComponentsWithVariables` are needed. I feel like what they are testing is already tested in previous tests...?

agarny committed:
Some minor cleaning up.

agarny commented:
At this stage, the only thing that is left for me (besides my "concerns" about unnecessary tests, which I am not going to hold against this PR) is to fix the math indentation issue, if possible (see https://github.com/cellml/libcellml/pull/483#discussion_r363150645).

agarny committed:
Make ClangFormat happy.

Not my preference, but ok whatever makes ClangFormat happy.

agarny committed:
Some minor cleaning up.

kerimoyle commented on src/model.cpp:
```diff
@@ -472,4 +577,208 @@ ModelPtr Model::clone() const
+    return rebasedMap;
+}
+
+void getComponentNames(const ComponentPtr &component, NameList &names)
```
Since this returns the component names, should it be consistent with function names elsewhere and ditch the "get"?

agarny replied:
Although it's an internal method, I agree. Done.

kerimoyle commented on src/model.cpp:
```diff
@@ -472,4 +577,208 @@ ModelPtr Model::clone() const
+}
+
+void Model::flatten()
+{
```
Again for consistency ... should this be flattenModel() (like validateModel and parseModel ?)

agarny replied:
Hmm, indeed. This means that we also need to rename `Model::clone()` to `Model::cloneModel()`. Done.

nickerso replied:
this is a method on `Model`, so I don't think it needs to have the model, that is by definition what is being flattened or cloned.

agarny replied:
Oops, I agree with @nickerso. For the `Validator` and `Parser` classes, it's a different story since we are indeed validating/parsin a model, not a validator/parser.

kerimoyle commented:
Do we expect users to be allowed to import two separate (and individually invalid) models and have them combine to something sensible?  For example, one model has all the units, another model has all the variables to which the units are linked by name.  Would we expect the combined model to report unlinked units or not?

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Parser: make sure that appended math is properly indented.

agarny committed:
Some minor cleaning up.

agarny committed:
Merge remote-tracking branch 'upstream/develop' into model_flattening_i386.

agarny committed:
Some minor cleaning up.

agarny committed:
Validator test: fixed a small problem in our invalidSimpleMathmlCellMLUnits test.

Indeed, the math  test string had been reformatted a bit, which now results in an additional error.

agarny commented:
> Do we expect users to be allowed to import two separate (and individually invalid) models and have them combine to something sensible? For example, one model has all the units, another model has all the variables to which the units are linked by name. Would we expect the combined model to report unlinked units or not?

I would say that there is nothing that prevents us from importing two separate (and individually invalid) models. Then, sure, this may result in unlinked units, but it shouldn't if you flatten your model (unless your combined model is also invalid).

agarny committed:
Get things to build on Windows and Linux.

agarny reviewed (approved):
Looks good to me apart from a few tests that I am not sure are needed. I guess I must be missing something, so not holding it against this PR.
